virtual_rainforest/core/grid.py

```diff
@@ -446,3 +452,132 @@ def populate_distances(
+            a given point can be empty - when the point falls in no cell - or of length
+            > 1 when a point falls on adjoining cell boundaries.
+        """
+
+        if (x_coords.ndim != 1) or (y_coords.ndim != 1):
+            raise ValueError("The x/y coordinate arrays are not 1 dimensional")
+
+        if x_coords.shape != y_coords.shape:
+            raise ValueError("The x/y coordinates are of unequal length")
+
+        # Get shapely points for the coordinates
+        xyp = [Point(x, y) for x, y in zip(x_coords, y_coords)]
+
+        # Map the Cell ID of each point - this is doing all pairs of points and cells,
+        # which is a computational choke point. Might be able to use a spatial search
+        # option, if this gets problematic, possibly including an STRTree in the grid
+        #    object https://shapely.readthedocs.io/en/latest/strtree.html
+
+        cell_map = [
+            [id for id, ply in zip(self.cell_id, self.polygons) if ply.intersects(pt)]
+            for pt in xyp
+        ]
+
+        return cell_map
+
+    def map_xy_to_cell_indexing(
+        self,
+        x_coords: np.ndarray,
+        y_coords: np.ndarray,
+        x_idx: Optional[np.ndarray],
+        y_idx: Optional[np.ndarray],
+        strict: Optional[bool] = True,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Returns indexing to map xy coordinates a single cell_id axis.
+
+        This function maps the provided one-dimensional set of x and y points onto the
+        grid (using `~virtual_rainforest.core.grid.Grid.map_xy_to_cell_id`) and then
+        checks that the mapped points uniquely identify all of the grid cells. If
+        `strict` is set to False, then extra points outside the grid are permitted.
+
+        The function then returns a pair of arrays that give indices on the original
+        x and y data to extract data along a single cell id axis. If specific x and y
+        indexing data is not provided, they are assumed to be simple sequences along
+        `x_coords` and `y_coords`.
+
+        Args:
+            cell_map: An existing mapping from points onto cell ids.
+            strict: Should points outside the grid be allowed
+
+        Returns:
+            A list showing the integer cell id for each pair of points. If strict is
+            False, the list element for points outside the grid is None.
+        """
+
+        # Get the coordinate mapping to cell ids
+        cell_map = self.map_xy_to_cell_id(x_coords=x_coords, y_coords=y_coords)
+
+        # Set indexing to sequence along coords if missing
+        if (x_idx is None) and (y_idx is None):
+            _x_idx = np.arange(x_coords.shape[0])
+            _y_idx = np.arange(y_coords.shape[0])
+        elif (x_idx is None) ^ (y_idx is None):  # Note: ^ is xor
+            raise ValueError("Only one of x/y indices provided.")
+        else:
+            _x_idx = x_idx  # type: ignore
+            _y_idx = y_idx  # type: ignore
```

Comment:
Two suggestions:

1. I'd do the check for the error condition first, then you don't need any `else`/`elif` clauses after it
2. I'd just assign the default values here to `x_idx` and `y_idx` directly rather than creating new variables. Fewer variables means less cognitive load for the person reading your code

Reply:
That's a `mypy` thing: `x_idx` has type `Optional[np.ndarray]`. This section of the code guarantees that it isn't `None` but later functions don't know that and  mypy objects. For example using `zip(cell_id_map, x_idx, y_idx)` results in:

"Argument 2 to "zip" has incompatible type "Optional[ndarray[Any, Any]]"; expected "Iterable[Any]"  [arg-type]mypy(error)"